src/main/java/org/commoncrawl/spark/util/CCIndex2FilenameParser.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.commoncrawl.spark.util;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class CCIndex2FilenameParser {
+    protected static final Pattern filenameAnalyzer = Pattern
+                .compile("^(?:common-crawl/)?crawl-data/([^/]+)/segments/([^/]+)/(crawldiagnostics|robotstxt|warc|wat|wet)/");
+
+                // crawl-data/CC-NEWS/2019/01/CC-NEWS-20190101042830-00057.warc.gz
+    protected static final Pattern newsFilenameAnalyzer = Pattern
+                .compile("^(?:common-crawl/)?crawl-data/CC-NEWS/(\\d+)/(\\d+)/CC-NEWS-(.+)\\.warc\\.gz");
+
+    // Class to encapsulate the extracted crawl, segment, and subset.
+    public static class FilenameParts {
+        public String crawl;
+        public String segment;
+        public String subset;
+    }
+
+    // Error class if we can't find the crawl, segment, and subset.
+    public static class FilenameParseError extends Exception {
+        public FilenameParseError(String message) {
+            super(message);
+        }
+    }
+
+    public static FilenameParts getParts(String filename) throws FilenameParseError {
+        FilenameParts parts = new FilenameParts();
+        Matcher m = filenameAnalyzer.matcher(filename);
+        if(m.find()){
+            parts.crawl = m.group(1);
+            parts.segment = m.group(2);
+            parts.subset = m.group(3);
+        } else {
+            Matcher newsParts = newsFilenameAnalyzer.matcher(filename);
+            if(!newsParts.find()){
+                throw new FilenameParseError("Filename not parseable (tried normal and news): " + filename);
+            }
+            parts.crawl = String.format("CC-NEWS-%s-%s", newsParts.group(1), newsParts.group(2));
+            parts.segment = newsParts.group(3);
```

Comment:
There is (now) no equivalent to the segment in the news crawl. I wouldn't use the WARC timestamp (time when the WARC was started) and its serial number as an equivalent: timestamps and serial number of the segments in the main crawl relate to time of fetch list generation. In the news crawl it's about the fetch time.

If segment is left away we would need to modify the schema for the news crawl because the "segment" is defined as *not nullable*. Maybe it's better to provide a different schema than trying to force every data set into the same scheme independent of its scope, collection method and tools. This was one insight of adapting the CDX-to-Parquet converter for the [end-of-term archive](https://eotarchive.org/). See [eot-index-schema.json](https://github.com/commoncrawl/cc-index-table/blob/eot-archive/src/main/resources/schema/eot-index-schema.json) and [EOTIndexTable.java](https://github.com/commoncrawl/cc-index-table/blob/eot-archive/src/main/java/org/commoncrawl/spark/EOTIndexTable.java).

Reply:
Yes, I thought much about this as well.

I did it this way for the following reasons:

- the redaction tools (and IMHO, all our tools) are going to be simpler if we're dealing with same index format.
- I think while news may have no natural "segment", I don't really see the harm with populating it with some potentially useful data. Perhaps I need to be aware of what the harm might be. (what do people even use that field for?)
- I also considered a single value ie: "0000000", but, per previous point, I think the way I did it is potentially a little more useful.

Perhaps it boils down to these options:

- Leave my solution (probably not)
- Populate only with serial number (seems a bit better based on what I think you're saying above)
- Populate with static value (I'm not a fan of useless data, but, I like it better than changing schema)
- New index format (makes many of our tools & future processes more complex "forever")

I'm going to ask a question below about news as well - that may be a part of the answer. See below.

Reply:
See the discussion below about the subset "news-warc".

> (what do people even use that field for?)

Only very few users use it, for example [Henry Thompson in "Improved methodology ..."](https://www.pure.ed.ac.uk/ws/portalfiles/portal/430071456/Improved_methodology_THOMPSON_DOA31012024_AFV_CC_BY.pdf). In order to use it you need to know what it stands for: one part (one 100th) of the fetch list with all data of one segment fetched together in the same time frame (as of now: 3 hours). We might explain it to our users, for example in the [schema](/commoncrawl/cc-index-table/blob/main/src/main/resources/schema/cc-index-schema-flat.json).